lib/plausible/ingestion/request.ex

```diff
@@ -149,42 +153,56 @@ defmodule Plausible.Ingestion.Request do
     Changeset.put_change(changeset, :hostname, sanitize_hostname(host))
   end
 
-  defp parse_props(%{} = request_body) do
-    raw_props =
-      request_body["m"] || request_body["meta"] || request_body["p"] || request_body["props"]
+  defp put_props(changeset, %{} = request_body) do
+    props =
+      (request_body["m"] || request_body["meta"] || request_body["p"] || request_body["props"])
+      |> decode_props_or_fallback()
+      |> Enum.reject(fn {_k, v} -> is_nil(v) || is_list(v) || is_map(v) || v == "" end)
+      |> Map.new()
 
-    case decode_raw_props(raw_props) do
-      {:ok, parsed_json} ->
-        parsed_json
-        |> Enum.filter(&valid_prop_value?/1)
-        |> Map.new()
-
-      _error ->
-        %{}
-    end
+    changeset
+    |> Changeset.put_change(:props, props)
+    |> validate_props()
   end
 
-  defp decode_raw_props(props) when is_map(props), do: {:ok, props}
-
-  defp decode_raw_props(raw_json) when is_binary(raw_json) do
-    case Jason.decode(raw_json) do
-      {:ok, parsed_props} when is_map(parsed_props) ->
-        {:ok, parsed_props}
-
-      _ ->
-        :not_a_map
+  defp decode_props_or_fallback(raw) do
+    with raw when is_binary(raw) <- raw,
+         {:ok, %{} = decoded} <- Jason.decode(raw) do
+      decoded
+    else
+      already_a_map when is_map(already_a_map) -> already_a_map
+      {:ok, _list_or_other} -> %{}
+      {:error, _decode_error} -> %{}
+      _any -> %{}
     end
   end
 
-  defp decode_raw_props(_), do: :bad_format
-
-  defp valid_prop_value?({key, value}) do
-    case {key, value} do
-      {_key, ""} -> false
-      {_key, nil} -> false
-      {_key, value} when is_list(value) -> false
-      {_key, value} when is_map(value) -> false
-      {_key, _value} -> true
+  @max_props 50
+  @max_prop_key_length 300
+  @max_prop_value_length 2000
+  defp validate_props(changeset) do
+    props = Changeset.get_field(changeset, :props)
+
+    props_too_large? =
+      Enum.any?(props, fn
+        {key, _value} when byte_size(key) > @max_prop_key_length -> true
+        {_key, value} when byte_size(value) > @max_prop_value_length -> true
+        _prop -> false
+      end)
+
+    cond do
+      Enum.count(props) > @max_props ->
+        Changeset.add_error(changeset, :props, "should not have more than #{@max_props} items")
+
+      props_too_large? ->
+        Changeset.add_error(
+          changeset,
+          :props,
+          "keys should have at most #{@max_prop_key_length} chars and values #{@max_prop_value_length} chars"
+        )
+
+      true ->
+        changeset
     end
```

Comment:
I think we can make it more compact by leveraging pattern matching and guards over `cond` and more efficient by only checking props length when the map size is valid. What do you think about:

```suggestion
defp validate_props(changeset) do
    case Changeset.get_field(changeset, :props) do
      props when map_size(props) > @max_props ->
        Changeset.add_error(changeset, :props, "should not have more than #{@max_props} items")

      props ->
        Enum.reduce_while(props, changeset, fn
          {key, value}, changeset
          when byte_size(key) > @max_prop_key_length or
                 byte_size(value) > @max_prop_value_length ->
            {:halt,
             Changeset.add_error(
               changeset,
               :props,
               "keys should have at most #{@max_prop_key_length} chars and values #{@max_prop_value_length} chars"
             )}

          _, changeset ->
            {:cont, changeset}
        end)
    end
```

Reply:
We also need to keep in mind that we're counting bytes here, not graphemes, so the "at most ... chars" error message might be confusing?

```
iex(1)> s = "我"
"我"
iex(2)> String.length(s)
1
iex(3)> byte_size(s)
3
```

Reply:
Good one! I replaced chars with bytes in the error message. Changed in 26ac00d8.